builtin/logical/database/path_roles_test.go

```diff
@@ -1087,6 +1087,91 @@ func TestBackend_StaticRole_Role_name_check(t *testing.T) {
 	}
 }
 
+// TestStaticRole_NewCredentialGeneration verifies that new
+// credentials are generated if a retried credential continues
+// to fail
+func TestStaticRole_NewCredentialGeneration(t *testing.T) {
+	ctx := context.Background()
+	b, storage, mockDB := getBackend(t)
+	defer b.Cleanup(ctx)
+	configureDBMount(t, storage)
+
+	roleName := "hashicorp"
+	createRole(t, b, storage, mockDB, "hashicorp")
+
+	t.Run("rotation failures should generate new password on retry", func(t *testing.T) {
+		// Fail to rotate the role
+		generateWALFromFailedRotation(t, b, storage, mockDB, roleName)
+
+		// Get WAL
+		walIDs := requireWALs(t, storage, 1)
+		wal, err := b.findStaticWAL(ctx, storage, walIDs[0])
+		if err != nil || wal == nil {
+			t.Fatal(err)
+		}
+
+		// Store password
+		initialPassword := wal.NewPassword
+
+		// Rotate role manually and fail again #1
+		generateWALFromFailedRotation(t, b, storage, mockDB, roleName)
+
+		// Ensure WAL is deleted since retrying initial password failed
+		requireWALs(t, storage, 0)
+
+		// Rotate role manually and fail again #2
+		generateWALFromFailedRotation(t, b, storage, mockDB, roleName)
```

Comment:
Do we need this call to `generateWALFromFailedRotation`? If I am following the test setup correctly, I think the previous 2 calls to `generateWALFromFailedRotation` should be sufficient.

Reply:
So here's the test flow in my head, for clarity. Let me know if there is an extra step here:
- Try rotating a password on L1104, and fail. This generates a failed password
- Try rotating password again on L1117, and fail. This will reuse the previous failing password. Since it fails again, the credential should be deleted after this failure step
- Try rotating and fail again. This ensures that since the WAL was deleted in previous step, a new WAL entry and credential were created. This now stores the new password in `secondRetryPassword`
- Finally, we rotate the role and succeed. This time, the failing password in step 4 was reused, and since it succeeded, was used as the credential. Since it succeeded, WAL entry is also deleted
- Last step is due diligence. We read the credentials to confirm that the password on a successful rotation was indeed `secondRetryPassword` from Step 4

In my mind this encapsulated the flow of multiple failing cases and then a success, and ensures the right expected credential is being used after rotations. LMK if you think any of these steps is unnecessary, thanks!

Reply:
My understanding of the change is that we generate a new password after the second failed attempt to rotate. Your test has 3 rotation failures. I think we want the test to assert that a new password is generated after the second failed attempt and/or the WAL has been deleted.

Reply:
Thanks for syncing up on this! Updated the test and removed the extra step